### PR TITLE
#233 support to decorate jpa repositories

### PR DESCRIPTION
**Pull request by remmeier**
JpaModule-based repositories provide the entire functionality out-of-the-box without any need for further implementation. To customize, implement actions, etc. it would be worthwhile to have decorators that wrap the original jpa repository.

Mechanism is applicable to any other kind of repository as well.

### CODE REVIEW

**Comment by remmeier** [comment]
configuration seems to be wrong, review is a good idea. But the details link to the coverage reports is missing.

**Comment by remmeier** [comment]
review welcomed :-) I will fix up the sonar coverage next to 90 to 95%.

**Comment by Ramblurr** [comment]
Will check this out, I'm very interested in JPA support.

As usual some documentation would be appreciated ;-)

File: katharsis-core/src/main/java/io/katharsis/module/ModuleRegistry.java [hunk 1]
```diff
@@ -390,65 +403,122 @@ public ResourceInformationBuilder getResourceInformationBuilder() {
...
+	}
 
-			ResourceEntry resourceEntry;
-			if (ClassUtils.getAnnotation(repository.getClass(), JsonApiResourceRepository.class).isPresent()) {
-				resourceEntry = new AnnotatedResourceEntry(repositoryInstanceBuilder);
-			}
-			else {
-				resourceEntry = new DirectResponseResourceEntry(repositoryInstanceBuilder);
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	private Object decorateRepository(Object repository) {
```
Comment (chb0github) [inline katharsis-core/src/main/java/io/katharsis/module/ModuleRegistry.java]: This function is

1. A bit too long
2. The use of `instanceof` is troubling
3. These conditional blocks are nearly identical

Applying a common interface to both of these repository type is the way to go.

File: katharsis-core/src/main/java/io/katharsis/repository/decorate/RepositoryDecoratorFactoryBase.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+import java.io.Serializable;
+
+import io.katharsis.repository.RelationshipRepositoryV2;
+import io.katharsis.repository.ResourceRepositoryV2;
+
+/**
+ * Base class for {@links RepositoryDecorator} implementations doing nothing.
+ */
+public class RepositoryDecoratorFactoryBase implements RepositoryDecoratorFactory {
```
Comment (chb0github) [inline katharsis-core/src/main/java/io/katharsis/repository/decorate/RepositoryDecoratorFactoryBase.java]: This class should probably be abstract

File: katharsis-core/src/test/java/io/katharsis/module/DecoratorTest.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import io.katharsis.repository.decorate.ResourceRepositoryDecoratorBase;
+import io.katharsis.resource.mock.models.Schedule;
+import io.katharsis.resource.mock.models.Task;
+import io.katharsis.resource.mock.repository.ScheduleRepository;
+
+public class DecoratorTest {
+
+	@SuppressWarnings({ "unchecked", "rawtypes" })
+	@Test
+	public void decoratedResourceRepositoryBase() {
```
Comment (chb0github) [inline katharsis-core/src/test/java/io/katharsis/module/DecoratorTest.java]: please prefix your test names with `test`

File: katharsis-rs/src/main/java/io/katharsis/rs/KatharsisFilter.java [hunk 4]
```diff
@@ -154,12 +154,17 @@ private void dispatchRequest(ContainerRequestContext requestContext) throws Exce
...
+       
         } catch (KatharsisMappableException e) {
+        	e.printStackTrace();
             // log error in KatharsisMappableException mapper.
             katharsisResponse = new KatharsisExceptionMapper().toErrorResponse(e);
         } catch (KatharsisMatchingException e) {
+        	e.printStackTrace();
         	LOGGER.warn("failed to process request", e);
             passToMethodMatcher = true;
+        } catch (Throwable e) {
```
Comment (chb0github) [inline katharsis-rs/src/main/java/io/katharsis/rs/KatharsisFilter.java]: no!

Why did you add this?

File: katharsis-core/src/test/java/io/katharsis/module/DecoratorTest.java (hunk 3 above)
Comment (remmeier) [inline katharsis-core/src/test/java/io/katharsis/module/DecoratorTest.java]: ok to be, different projects, different opinions about that...

File: katharsis-rs/src/main/java/io/katharsis/rs/KatharsisFilter.java (hunk 4 above)
Comment (remmeier) [inline katharsis-rs/src/main/java/io/katharsis/rs/KatharsisFilter.java]: KatharsisMappableException handling might also has some room for optimization.

Comment (remmeier) [inline katharsis-rs/src/main/java/io/katharsis/rs/KatharsisFilter.java]: sonar setup is not workign anymore? usually it taskes care of this kind of things.

**Comment by remmeier** [comment]
added the changes, thx!
